pkg/sbom/scanner/scanner.go

```diff
@@ -0,0 +1,177 @@
+		return nil
+	default:
+		return fmt.Errorf("collector queue for '%s' is full", collectorName)
+	}
+}
+
+func (s *Scanner) enoughDiskSpace(opts sbom.ScanOptions) error {
+	if !opts.CheckDiskUsage {
+		return nil
+	}
+
+	usage, err := s.disk.GetUsage("/")
+	if err != nil {
+		return err
+	}
+
+	if usage.Available < opts.MinAvailableDisk {
+		return fmt.Errorf("not enough disk space to safely collect sbom, %d available, %d required", usage.Available, opts.MinAvailableDisk)
+	}
+
+	return nil
+}
+
+func (s *Scanner) Start(ctx context.Context) {
+	if s.running {
+		return
+	}
+
+	go func() {
+		s.running = true
+		defer func() { s.running = false }()
+
+		for {
+			select {
+			// We don't want to keep scanning if image channel is not empty but context is expired
+			case <-ctx.Done():
+				return
+
+			case request, ok := <-s.scanQueue:
+				// Channel has been closed we should exit
+				if !ok {
+					return
+				}
+
+				sbomAttempts.Inc(request.Collector(), request.Type())
+
+				collector := request.collector
+				if err := s.enoughDiskSpace(request.opts); err != nil {
+					sbomFailures.Inc(request.Collector(), request.Type(), "disk_space")
+					log.Errorf("An error occurred while checking current disk usage: %s", err)
+					continue
+				}
+
+				scanContext, cancel := context.WithTimeout(ctx, scanTimeout)
+				createdAt := time.Now()
+				report, err := collector.Scan(scanContext, request.ScanRequest, request.opts)
+				generationDuration := time.Since(createdAt)
+				cancel()
+				if err != nil {
+					sbomFailures.Inc(request.Collector(), request.Type(), "scan")
+					log.Errorf("An error occurred while generating SBOM: %s", err)
+					continue
+				}
+
+				telemetry.SBOMGenerationDuration.Observe(generationDuration.Seconds())
+
+				select {
+				case request.ch <- sbom.ScanResult{
+					Report:    report,
+					CreatedAt: createdAt,
+					Duration:  generationDuration,
+				}:
+				default:
+					log.Errorf("Failed to push scanner result into channel")
+				}
+
+				if request.opts.WaitAfter != 0 {
+					t := time.NewTimer(request.opts.WaitAfter)
+					select {
+					case <-ctx.Done():
+					case <-t.C:
+					}
+					t.Stop()
+				}
+			}
+		}
+	}()
+}
+
+// NewScanner creates a new SBOM scanner. Call Start to start the store and its
+// collectors.
+func NewScanner(cfg config.Config) *Scanner {
+	return &Scanner{
+		scanQueue: make(chan scanRequest, 500),
+		disk:      filesystem.NewDisk(),
+	}
+}
+
+// CreateGlobalScanner creates a SBOM scanner, sets it as the default
+// global one, and returns it. Start() needs to be called before any data
+// collection happens.
+func CreateGlobalScanner(cfg config.Config) (*Scanner, error) {
+	if !cfg.GetBool("sbom.enabled") && !cfg.GetBool("container_image_collection.sbom.enabled") {
+		return nil, nil
+	}
+
+	if globalScanner != nil {
+		return nil, errors.New("global SBOM scanner already set, should only happen once")
+	}
+
+	for name, collector := range collectors.Collectors {
```

Comment:
When `CreateGlobalScanner` is called, this list is empty as nobody imported all scanners. Currently the scanners are lazily added on usage when someone needs the `ScanRequest`.
I see two options:
- Having a `import _`, like `pkg/workloadmeta/collectors/collectors.go`
- Removing the `Init()` requirements by using global Config.